### PR TITLE
Issue #340 - create application/project events for audit

### PR DESCRIPTION
**Pull request by alexmt**
PR implements following changes:
- app controller, app/project api services produce k8s events on every app/project change
- api GET /applicaitons/<app-name>/events now return application events is resource name is not specified

PR has one questionable change: k8s event don't have suitable field to store username. I've chosen to use `ReportingInstance` field for username. All fields seems equally bad for this purpose. Please let me know if you have better idea.

### CODE REVIEW

**Comment by alexmt** [comment]
It make sense to me. Updated.

**Review by merenbach — APPROVED** [review]

File: server/project/project.go [hunk 1]
```diff
@@ -14,22 +14,27 @@ import (
...
 	appclientset appclientset.Interface
+	auditLogger  *argo.AuditLogger
 	projectLock  *util.KeyLock
 }
 
 // NewServer returns a new instance of the Project service
-func NewServer(ns string, appclientset appclientset.Interface, enf *rbac.Enforcer, projectLock *util.KeyLock) *Server {
-	return &Server{enf: enf, appclientset: appclientset, ns: ns, projectLock: projectLock}
+func NewServer(ns string, kubeclientset kubernetes.Interface, appclientset appclientset.Interface, enf *rbac.Enforcer, projectLock *util.KeyLock) *Server {
+	auditLogger := argo.NewAuditLogger(ns, kubeclientset, "argo-server")
```
[the remark this answers is not in the document]
Comment (alexmt) [reply server/project/project.go]: done

File: util/session/sessionmanager.go [hunk 2]
```diff
@@ -174,6 +174,19 @@ func (mgr *SessionManager) VerifyToken(tokenString string) (jwt.Claims, error) {
...
 }
 
+func Username(ctx context.Context) string {
+	if claims, ok := ctx.Value("claims").(*jwt.MapClaims); ok {
+		mapClaims := *claims
+		switch mapClaims["iss"] {
+		case SessionManagerClaimsIssuer:
+			return mapClaims["sub"].(string)
+		default:
+			return mapClaims["email"].(string)
```
[the remark this answers is not in the document]
Comment (alexmt) [reply util/session/sessionmanager.go]: fixed.

**Comment by alexmt** [comment]
Thanks for review @jessesuen . PTAL

File: server/application/application.go [hunk 3]
```diff
@@ -66,6 +68,7 @@ func NewServer(
 		appComparator: controller.NewAppStateManager(db, appclientset, repoClientset, namespace),
 		enf:           enf,
 		projectLock:   projectLock,
+		auditLogger:   argo.NewAuditLogger(namespace, kubeclientset, "argo-server"),
```
Comment (jessesuen) [inline server/application/application.go]: argocd-server

**Comment by jessesuen** [comment]
One more change to use argocd-server instead of argo-server.

I don't think it needs to be addressed in this change, but can you file a new issue for us to use constants for things like `terminateop`, `sync`, `rollback`, `update`, etc... All the ones that we share with the RBAC policy.csv which we are currently hard coding as strings.

**Review by jessesuen — APPROVED** [review]

**Pull request merged**